### ISSUE TITLE
Unable to compile workset iteration, URGENT

### ISSUE THREAD

**Issue by sscdotopen** [stratosphere/stratosphere#142]
Hi, I tried to create a simple incrementally iterative algorithm for the homework for the course here in Trento.

Unfortunately, I can't get the compilation of workset iterations to run.

I get an exception saying that I try to change the partition in the Match updating the solutionset (which I think I don't do :))

Could anyone have a look at the code? Its very urgent as I have to hand out the assignment this week!

Code is here: https://github.com/stratosphere/graphmining-tutorial/blob/master/src/main/java/de/tuberlin/dima/aim3/graphmining/chainletter/ChainLetter.java

```
Exception in thread "main" eu.stratosphere.pact.compiler.CompilerException: No viable strategies for solution set delta found during plan enumeration. Possible reason: Partitioning is not preserved when matching with the solution set. Missing annotation?
    at eu.stratosphere.pact.compiler.plan.WorksetIterationNode.instantiate(WorksetIterationNode.java:303)
    at eu.stratosphere.pact.compiler.plan.TwoInputNode.addLocalCandidates(TwoInputNode.java:534)
    at eu.stratosphere.pact.compiler.plan.TwoInputNode.getAlternativePlans(TwoInputNode.java:465)
    at eu.stratosphere.pact.compiler.plan.DataSinkNode.getAlternativePlans(DataSinkNode.java:251)
    at eu.stratosphere.pact.compiler.PactCompiler.compile(PactCompiler.java:712)
    at eu.stratosphere.pact.compiler.PactCompiler.compile(PactCompiler.java:547)
    at eu.stratosphere.pact.client.LocalExecutor.executePlan(LocalExecutor.java:80)
    at eu.stratosphere.pact.client.LocalExecutor.execute(LocalExecutor.java:137)
```

**Comment by StephanEwen** [comment]
Hey!

I think the issue is exactly as the error message says: You are missing an annotation that tells the optimizer that the keys are preserved.

Please add a @ConstantFieldsFirst to the class containing the MatchStub and try it again!

**Comment by sscdotopen** [comment]
Why do I have to add this annotation? If the Match changes the partitioning, our iteration implementation doesn't work anyways, right?

Adding the annotation gives me another error:

```
Exception in thread "main" eu.stratosphere.pact.compiler.CompilerException: An error occurred while translating the optimized plan to a nephele JobGraph: An error occurred while translating the optimized plan to a nephele JobGraph: null
    at eu.stratosphere.pact.compiler.plantranslate.NepheleJobGraphGenerator.postVisit(NepheleJobGraphGenerator.java:555)
    at eu.stratosphere.pact.compiler.plantranslate.NepheleJobGraphGenerator.postVisit(NepheleJobGraphGenerator.java:96)
    at eu.stratosphere.pact.compiler.plan.candidate.DualInputPlanNode.accept(DualInputPlanNode.java:222)
    at eu.stratosphere.pact.compiler.plan.candidate.SingleInputPlanNode.accept(SingleInputPlanNode.java:141)
    at eu.stratosphere.pact.compiler.plan.candidate.OptimizedPlan.accept(OptimizedPlan.java:161)
    at eu.stratosphere.pact.compiler.plantranslate.NepheleJobGraphGenerator.compileJobGraph(NepheleJobGraphGenerator.java:165)
    at eu.stratosphere.pact.client.LocalExecutor.executePlan(LocalExecutor.java:83)
    at eu.stratosphere.pact.client.LocalExecutor.execute(LocalExecutor.java:137)
    at de.tuberlin.dima.aim3.graphmining.chainletter.ChainLetter.main(ChainLetter.java:126)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:120)
Caused by: eu.stratosphere.pact.compiler.CompilerException: An error occurred while translating the optimized plan to a nephele JobGraph: null
    at eu.stratosphere.pact.compiler.plantranslate.NepheleJobGraphGenerator.postVisit(NepheleJobGraphGenerator.java:555)
    at eu.stratosphere.pact.compiler.plantranslate.NepheleJobGraphGenerator.postVisit(NepheleJobGraphGenerator.java:96)
    at eu.stratosphere.pact.compiler.plan.candidate.SolutionSetPlanNode.accept(SolutionSetPlanNode.java:76)
    at eu.stratosphere.pact.compiler.plan.candidate.DualInputPlanNode.accept(DualInputPlanNode.java:220)
    at eu.stratosphere.pact.compiler.plan.candidate.WorksetIterationPlanNode.acceptForStepFunction(WorksetIterationPlanNode.java:175)
    at eu.stratosphere.pact.compiler.plantranslate.NepheleJobGraphGenerator.postVisit(NepheleJobGraphGenerator.java:389)
    ... 13 more
Caused by: java.lang.NullPointerException
    at eu.stratosphere.pact.compiler.plantranslate.NepheleJobGraphGenerator.postVisit(NepheleJobGraphGenerator.java:450)
    ... 18 more
```

**Comment by StephanEwen** [comment]
That is exactly the point. If the match changes the partitioning, the current code path does not work. It would need to add an extra partitioning step and update the solution set elements deferred ( #21 ). Since that has not been implemented, yet, the compiler needs to reject the job.

What are suggestions to handle this until the deferred index updates are implemented?

I'll look into the other bug...

**Comment by sscdotopen** [comment]
I'm just saying I found it not user-friendly to have to use the annotation. It doesn't add any value to my code and prevented me from using the Workset Iteration :)

**Comment by StephanEwen** [comment]
The system can currently not handle situations where the matcher with the solution set changes the key. We can either say we accept all jobs, optimistically assuming they do not change the key, and potentially produce erroneous results or error reports, or we reject these jobs and only accept the ones where we are sure (the ones with the annotation). We currently does the later, but we can change that.

**Comment by sscdotopen** [comment]
I think that what is going to happen is that users will simply add the annotation to their code even if their code does not adhere the rules :)

**Comment by StephanEwen** [comment]
Okay, so we need to get the general code path working as fast as possible. Then the annotation is a performance thing only, and like all semantic assertions, will cause incorrect behavior if the code actually violates it.

**Comment by StephanEwen** [comment]
The bug is fixed in 0e53fe573f5bbbd6f33647f746d99fe70153ece7

**Comment by aljoscha** [comment]
@StephanEwen By the way, if you refer to an issue in your commit message or pull request the issue is closed automatically with a link to the commit that closed it. See here: https://help.github.com/articles/closing-issues-via-commit-messages
